Closes #6012: Run the FxA extension on non-prod servers

eoger opened the pull request:
Thanks to `GeckoPort.senderUrl`  exposing the content URL, we can now assert in `FxaWebChannelFeature` that a WebChannel message comes from the right content page.
This allows us to inject the extension content script in more FxA server instances while ensuring that a non-prod instance can mess up a prod-instance account data.
 
---
<!-- Text above this line will be added to the commit once "bors" merges this PRR -->

Also fixes https://github.com/mozilla/application-services/issues/2795.

### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [x] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- [x] **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

eoger commented:
Not sure who's the right person to review on a-c side, a-s side @rfk since you were having security concerns in https://github.com/mozilla-mobile/android-components/issues/6012#issuecomment-589341798.

rfk commented on components/feature/accounts/src/main/java/mozilla/components/feature/accounts/FxaWebChannelFeature.kt:
```diff
@@ -98,9 +101,19 @@ class FxaWebChannelFeature(
+        @SuppressWarnings("ComplexMethod")
         override fun onPortMessage(message: Any, port: Port) {
+            val senderOrigin = URL(port.senderUrl()).host
+            val expectedOrigin = URL(serverConfig.contentUrl).host
```
I'm not super familiar with `java.net.URL` but a quick read of the docs suggests that `host` may not be sufficient here, because it may not include e.g. the port number or URL scheme. I've seen enough origin-matching bugs in my time to be very wary of getting this right.  I'd like to see this check factored out into some sort of `isSameOrigin()` function and have tests for at least the following cases:

* `isSameOrigin("https://example.com", https://example.com") -> true` (simple match)
* `isSameOrigin("https://example.com/foo", https://example.com/bar") -> true` (matches ignoring path)
* `isSameOrigin("https://example.com", https://example.com:443") -> true` (default port for scheme)
* `isSameOrigin("https://example.com:8000", https://example.com:8000") -> true` (non-default port match)
* `isSameOrigin("https://example.com", https://example.net") -> false` (host mismatch)
* `isSameOrigin("https://example.com", http://example.com") -> false` (scheme mismatch)
* `isSameOrigin("https://example.com", https://example.com:444") -> false` (non-default port mismatch)
* `isSameOrigin("https://example.com:444", https://example.com:445") -> false` (explicit port mismatch)
* `isSameOrigin("https://example.com", http://example.com:443") -> false` (scheme mismatch despite explicit port)
* `isSameOrigin("https://example.com:8000", http://example.com:8000") -> true` (scheme mismatch despite non-default port)

I had a bit of a look around but wasn't able to find an easy way to do this ([this package](https://kotlinlang.org/api/latest/jvm/stdlib/org.w3c.dom.url/) provides a `URL` class with an `origin` property, but looks like it's only when targetting javascript?). Absent something better, I suggest something like (pardon my pseudocode):

```
def canonicalizeOrigin(v):
  u = URL(v)
  c = URL(u.protocol, u.host, u.port == -1 ? u.defaultPort : u.port, "")
  return c.toString()

def isSameOrigin(v1, v2):
    return canonicalizeOrigin(v1) == canonicalizeOrigin(v2)
```

It's not very efficient, but it's easy to understand.

I also think it would be valuable to add some belt-and-braces security checks for things we would never expect to se in the wild:

* Reject any messages received from a non-https non-localhost URL
* Reject any messages received with from a URL with non-empty userinfo component

rfk commented on components/feature/accounts/src/test/java/mozilla/components/feature/accounts/FxaWebChannelFeatureTest.kt:
```diff
@@ -144,17 +150,61 @@ class FxaWebChannelFeatureTest {
+        whenever(sessionManager.getOrCreateEngineSession(session)).thenReturn(engineSession)
+        whenever(port.engineSession).thenReturn(engineSession)
+        whenever(port.senderUrl()).thenReturn("https://bar.foo/email")
+        whenever(serverConfig.contentUrl).thenReturn("https://foo.bar")
```
As mentioned above, I think we should test a lot more mis-matched combinations here, but perhaps we can do that in a way that only tests the matching logic rather than having to spin up all this infra and actually send a request.

eoger commented on components/feature/accounts/src/main/assets/extensions/fxawebchannel/manifest.json:
```diff
@@ -4,7 +4,12 @@
   "content_scripts": [
     {
-      "matches": ["https://accounts.firefox.com/*"],
+      "matches": [
```
I should also add `latest` here.

eoger commented:
I have addressed @rfk's good comments, r? @csadilek

csadilek commented on docs/changelog.md:
```diff
@@ -12,6 +12,14 @@ permalink: /changelog/
+
+* **feature-accounts**
+  * It should now be possible to log-in on stable, stage and china FxA servers using a WebChannel flow.
+  * `FxaWebChannelFeature` takes a `ServerConfig` as 4th parameter to ensure incoming WebChannel messages
```
Let's mark this as a breaking change (there are examples in this files). We need to put up PRs to Fenix and RB in anticipation of this landing too, if you don't mind. Otherwise we'll break master.

csadilek commented on components/feature/accounts/src/main/java/mozilla/components/feature/accounts/FxaWebChannelFeature.kt:
```diff
@@ -299,5 +309,48 @@ class FxaWebChannelFeature(
+         * Compares 2 URLs and returns true if they have the same origin,
+         * which means: same protocol, same host, same port.
+         */
+        private fun isSameOrigin(a: String, b: String): Boolean {
```
Nice! Let's move this to a `String` extension function similar to https://github.com/mozilla-mobile/android-components/blob/master/components/support/ktx/src/main/java/mozilla/components/support/ktx/kotlin/String.kt#L105

Then we can also move the extensive test case over to StringUtilsTest. You can then also invoke this as `senderOrigin.isSameOriginAs(expectedOrigin)`.

eoger commented:
- Opened https://github.com/mozilla-mobile/fenix/pull/9151 for Fenix, r-b doesn't have the webchannel feature.
- Addressed your comments

Thanks @csadilek! If you're happy with it this is ready to land.

eoger commented:
Rebased and ready to land, let me know how you want to proceed.

csadilek commented:
bors r+

csadilek commented:
@rfk can you approve this PR to make bors happy :)

csadilek commented:
bors r+
